versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreIndexImpl.java

```diff
@@ -428,44 +442,49 @@ public int size() {
 
   @Override
   public int estimatedSerializedSize() {
-    return originalSerializedSize + estimatedSerializedSizeDiff + 1;
+    return originalSerializedSize + estimatedSerializedSizeDiff;
   }
 
   @Override
   public @Nonnull @jakarta.annotation.Nonnull ByteString serialize() {
-    ByteBuffer target = ByteBuffer.allocate(estimatedSerializedSize());
+    ByteBuffer target;
 
-    // Serialized segment index version
-    if (SERIALIZE_VERSION >= CURRENT_STORE_INDEX_VERSION) {
-      target.put((byte) 2);
-      putVarInt(target, elementCount());
-    } else {
-      target.put((byte) 1);
-    }
+    if (serialized == null || modified) {
+      target = ByteBuffer.allocate(estimatedSerializedSize());
 
-    ByteBuffer previousKey = null;
+      // Serialized segment index version
+      if (SERIALIZE_VERSION >= CURRENT_STORE_INDEX_VERSION) {
+        target.put((byte) 2);
+        putVarInt(target, elementCount());
+      } else {
+        target.put((byte) 1);
+      }
 
-    // This buffer's backs the currently serialized key - use with care!
-    // Having this "singleton" instance massively reduces GC alloc/churn rate.
-    ByteBuffer serializationBuffer = newKeyBuffer();
+      ByteBuffer previousKey = null;
 
-    @SuppressWarnings("UnnecessaryLocalVariable")
-    ElementSerializer<V> ser = serializer;
+      // This buffer's backs the currently serialized key - use with care!
+      // Having this "singleton" instance massively reduces GC alloc/churn rate.
 
-    for (StoreIndexElement<V> el : this) {
-      StoreKey key = el.key();
+      @SuppressWarnings("UnnecessaryLocalVariable")
+      ElementSerializer<V> ser = serializer;
 
-      previousKey = serializeKey(previousKey, key, target, serializationBuffer);
+      List<StoreIndexElement<V>> elements = this.elements;
+      for (int i = 0; i < elements.size(); i++) {
+        StoreIndexElement<V> el = elements.get(i);
+        StoreKey key = el.key();
+        previousKey = serializeKey(previousKey, key, target, scratchKeyBuffer);
+        el.serializeContent(ser, target);
+      }
 
-      ser.serialize(el.content(), target);
+      target.flip();
+    } else {
+      target = serialized.position(0).limit(originalSerializedSize);
     }
 
-    target.flip();
-
     return unsafeWrap(target);
   }
 
-  private ByteBuffer serializeKey(
+  private static ByteBuffer serializeKey(
       ByteBuffer previousKey, StoreKey key, ByteBuffer target, ByteBuffer serializationBuffer) {
     ByteBuffer keyBuf = key.serialize(serializationBuffer);
```

Comment:
Not related to this PR, but I noticed that `serializationBuffer` has a capacity of `MAX_KEY_BYTES`, and yet `StoreKey.serialize` puts the whole key plus 2 zero bytes in it. Isn't this effectively limiting the max key length to `MAX_KEY_BYTES` - 2?

Reply:
Yes, that's true.

Reply:
`ContentKey.MAX_LENGTH=500` though